stdlib/condition.mli

```diff
@@ -13,41 +13,112 @@
-   thread has finished doing something: the former thread 'waits' on the
-   condition variable, the latter thread 'signals' the condition when it
-   is done. Condition variables should always be protected by a mutex.
-   The typical use is (if [D] is a shared data structure, [m] its mutex,
-   and [c] is a condition variable):
+(**Condition variables.
+
+   Condition variables are useful when several threads wish to access a
+   shared data structure that is protected by a mutex (a mutual exclusion
+   lock).
+
+   A condition variable is a {i communication channel}. On the receiver
+   side, one or more threads can indicate that they wish to {i wait}
+   for a certain property to become true. On the sender side, a thread
+   can {i signal} that this property has become true, causing one (or
+   more) waiting threads to be woken up.
+
+   For instance, in the implementation of a queue data structure, if a
+   thread that wishes to extract an element finds that the queue is
+   currently empty, then this thread waits for the queue to become
+   nonempty. A thread that inserts an element into the queue signals
+   that the queue is nonempty. A condition variable is used for this
+   purpose. This communication channel conveys the information that
+   the property "the queue is nonempty" is true, or (more accurately)
+   may be true.
+
+   In short, a condition variable [c] is used to convey the information
+   that a certain property [P] about a shared data structure [D],
+   protected by a mutex [m], may be true.
+
+   With a condition variable [c], exactly one mutex [m] is associated.
+   This association is implicit: the mutex [m] is not explicitly passed
+   as an argument to {!create}. It is up to the programmer to know, for
+   each condition variable [c], which is the associated mutex [m].
+
+   With a mutex [m], several condition variables can be associated.
+   For instance, in a bounded queue, one condition variable could
+   be used to indicate that the queue is nonempty, and another condition
+   variable could be used to indicate that the queue is not full.
+
+   With a condition variable [c], exactly one logical property [P]
+   should be associated. Examples of such properties
+   include "the queue is nonempty" and "the queue is not full".
+   It is up to the programmer to keep track, for each condition
+   variable, of the corresponding property [P].
+   A signal should be sent on the condition variable [c]
+   only when the property [P] is definitely true.
+   When a signal is received, however, one cannot assume that [P] is
+   definitely true; one must explicitly test whether [P] is true.
+   The reason is that,
+   between the moment when the signal is sent
+   and the moment when a waiting thread receives the signal
+   and is scheduled,
+   the property [P] may be falsified by some other thread
+   that is able to acquire the mutex [m] and alter the data structure [D].
```

Comment:
Program logic can, in some situations (single writer, single reader,  for instance) guarantee that [P} is true when twaiting thread wakes up. However, there is another reason for checking [P]  a waiting thread can wake up for no reason (_spurious wakeup_).

Reply:
OK, I have added a mention of spurious wakeups.